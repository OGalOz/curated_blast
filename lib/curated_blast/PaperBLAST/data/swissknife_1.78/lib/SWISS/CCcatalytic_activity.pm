package SWISS::CCcatalytic_activity;

use vars qw($AUTOLOAD @ISA %fields);

use Carp;
use strict;
use SWISS::TextFunc;
use SWISS::BaseClass;
use SWISS::CC;


BEGIN {
    @ISA = ( 'SWISS::BaseClass' );

    %fields = (
        activity   => undef, # {  rhea => "", ec => "", reaction => "", participants => [], ev => "" }
        directions => undef, # [ { txt => "", rhea => "", ev => "" } ]
        form       => undef,
        _old       => undef  # [ [ txt, ev ], ]
    );
}


sub new {
    my $ref   = shift;
    my $class = ref($ref) || $ref;
    my $self  = new SWISS::BaseClass;
    $self->rebless($class);
    return $self;
}


sub fromText {
    my $class   = shift;
    my $textRef = shift;
    my $self    = new SWISS::CCcatalytic_activity;
    my $text    = $$textRef;
    $self->initialize();
    $text =~ s/ *-!- CATALYTIC ACTIVITY: +//;
    $text =~ s/ {2,}/ /g;

    if ( $text =~ /Reaction=/ ) { # new format (rhea/ec & reaction)
        my $core = {}; my $iscore = 0;
        foreach my $elem ( split /; /, $text ) {
            $elem =~ s/;$//;
            my ( $key, $val ) = ( "", "" );
            ( $key, $val ) = $elem =~ /^(\w+)=(.+)$/;
            if ( $key eq "Reaction" ) {
                $iscore = 1;
                $core->{reaction} = $val;
            }
            elsif ( $key eq "Xref" ) {
                my @ids = split /, /, $val;
                ( my $rhea_id = shift @ids ) =~ s/^Rhea://;
                if ( $iscore ) { # core (reaction)
                    $core->{rhea} = $rhea_id;
                    $core->{participants} = \@ids;
                }
                else           { # direction
                    my $dir = $self->{ directions }->[ -1 ] || { 'txt' => undef, 'ev' => undef };
                    $dir->{rhea} = $rhea_id;
                }
            }
            elsif ( $key eq "EC" ) {
                $core->{ec} = $val;
            }
            elsif ( $key eq "Evidence" ) {
                if ( $iscore ) { $core->{ev} = $val } # core (reaction)
                else           { # direction
                    my $dir = $self->{ directions }->[ -1 ] || { 'txt' => undef, 'rhea' => undef };
                    $dir->{ev} = $val;
                }
            }
            elsif ( $key eq "PhysiologicalDirection" ) {
                $iscore = 0;
                push @{ $self->{directions} }, { 'txt' => $val, 'rhea' => undef, 'ev' => undef };
            }
        }
        $self->{activity} = $core;
    }
    else { # old unstructured format
        my $has_new_style_ev = $text =~ /ECO:/;
        $self->{_old} = SWISS::CC::parse2Blocks( $text );
        # set evidenceTags 4 compatibility with existing ...EvidenceTag methods in BaseClass! + old tests (fixme: clean/remove that mess!?)
        # but the real evidences are now within blocks (2nd field of block array)
        my @evs = map { split /, ?/, $_->[1] } grep { $_->[1] } @{ $self->{_old} };
        my $ev  = $has_new_style_ev ? " {" . join( ", ", @evs ) . "}" : "{" . join( ",", @evs ) . "}";
        $self->evidenceTags( $ev ) if @evs;
    }

    $self->{ _dirty } = 0;
    return $self;
}


sub topic {
    return "CATALYTIC ACTIVITY";
}


sub activity_rhea { # get (or set) rhea (id without "Rhea:" prefix, but with RHEA:) e.g. (RHEA:18037)
    my ( $self, $value, $ev ) = @_;

    if ( !$self->{_old} ) { # only works with new (rhea/ec + reaction) format
        if ( defined $value ) {
            my $new_ev = $ev ? $ev : $self->{ activity }->{ev};
            $self->{activity}->{rhea} = $value;
            $self->{activity}->{ev}   = $new_ev;
        }

        return $self->{activity} ? $self->{activity}->{rhea} : "";
    }
    else { return "" }
}

sub activity_reaction { # get (or set) reaction
    my ( $self, $value, $ev_only_for_old_format ) = @_;

    if ( !$self->{_old} ) { # new (rhea/ec + reaction) format
        if ( defined $value ) {
            $self->{activity}->{reaction} = $value;
        }

        return $self->{activity} ? $self->{activity}->{reaction} : "";
    }
    else { # old format
        if ( defined $value ) {
            my $new_ev = $ev_only_for_old_format ? $ev_only_for_old_format : " {" . join( ", ", map { split /, ?/, $_->[1] } grep { $_->[1] } @{ $self->{_old} } ) . "}";
            $self->evidenceTags( $new_ev );
            $self->{_old} = [ [ $value, $new_ev ] ];
        } # -> ! not tested
        return join( ". ", map { $_->[0] } @{ $self->{_old} } )."."
    }
}

sub activity_ec { # get (or set) ec
    my ( $self, $value, $ev ) = @_;

    if ( !$self->{_old} ) { # only works with new (rhea/ec + reaction) format
        if ( defined $value ) {
            my $new_ev = $ev ? $ev : $self->{ activity }->{ev};
            $self->{activity}->{ec} = $value;
            $self->{activity}->{ev} = $new_ev;
        }

        return $self->{activity} ? $self->{activity}->{ec} : "";
    }
    else { return "" }
}

sub activity_participant {
    my ( $self, $values ) = @_;

    if ( !$self->{_old} ) { # only works with new (rhea/ec + reaction) format
        if (defined $values) {
            $self->{activity}->{participants} = $values;
        }
        return $self->{activity} ? $self->{activity}->{participants} : [];
    }
    else { return [] }
}

sub directions { # get (or set) direction(s) structures
    my ( $self, $values ) = @_;

    if ( !$self->{_old} ) { # only works with new (rhea/ec + reaction) format
        if ( defined $values && ref $values eq "ARRAY") {
            $self->{directions} = $values;
        }
        return $self->{directions};
    }
    else { return [] }
}


sub toString {
    my ( $self ) = @_;

    my $activity_txt = $self->_activity_to_text();

    my $text = "CC   -!- CATALYTIC ACTIVITY:";
    if ( !$self->{_old} ) { # new (rhea/ec, reaction) format
        #my $form = $self->{form} ? " $self->{form}:" : "";
        $text .= "\n";
        $text .= SWISS::TextFunc->wrapOn( 'CC       ',"CC         ", $SWISS::TextFunc::lineLength, $activity_txt );
        map {
            my $direction_txt = _direction_struct_to_text( $_ );
            $text .= SWISS::TextFunc->wrapOn( 'CC       ',"CC         ", $SWISS::TextFunc::lineLength, $direction_txt );
        } @{ $self->{directions} };
    }
    else { # old unstructured format
        $text .= " ".$activity_txt;
        $text = SWISS::TextFunc->wrapOn('',"CC       ", $SWISS::TextFunc::lineLength, $text );
    }

    return $text;
}

sub _activity_to_text {
    my ( $self, $no_ev ) = @_;
    if ( !$self->{_old} ) { # new structured format
        my $activity     = $self->{ activity };
        my $activity_txt = "Reaction=$activity->{reaction};";
        if ( $activity->{rhea} ) {
            $activity_txt .= " Xref=Rhea:$activity->{rhea}";
            if ( $activity->{participants} )  { $activity_txt .= ", " . join( ", ", @{ $activity->{participants} } ) }
            $activity_txt .= ";";
        }
        if ( $activity->{ec} )            { $activity_txt .= " EC=$activity->{ec};" }
        if ( $activity->{ev} && !$no_ev ) { $activity_txt .= " Evidence=$activity->{ev};" }
        return $activity_txt;
    }
    else { # old format
        return ( $no_ev ? join( ". ", map { $_->[0] } @{ $self->{_old} } )."." : SWISS::CC::blocks2String( $self->{_old}, $self->evidenceTags )."." );
    }
}
sub _direction_struct_to_text {
    my ( $direction, $no_ev ) = @_;
    my $direction_txt = "";
    if ( $direction->{txt} )           { $direction_txt .= "PhysiologicalDirection=$direction->{txt};" }
    if ( $direction->{rhea} )          { $direction_txt .= " Xref=Rhea:$direction->{rhea};" }
    if ( $direction->{ev} && !$no_ev ) { $direction_txt .= " Evidence=$direction->{ev};" }
    return $direction_txt;
}

sub comment { # unwrapped content (without ev)
    my ($self, $with_ev) = @_;

    my $activity_txt  = $self->_activity_to_text (  "no_ev=true" );
    my $direction_txt = "";
    map { $direction_txt .= ( $direction_txt ? " " : "" ) . _direction_struct_to_text( $_ , "no_ev=true" ); } @{ $self->{directions} };
    return $activity_txt . ( $direction_txt ? " $direction_txt" : "" );
}


#sub sort {
#    my $self = shift;
#    $self->{ list } = sort { lc( $a->[0] ) cmp lc( $b->[0] ) } @{ $self->{ list } };
#}


1;

__END__

=head1 Name

SWISS::CCcatalytic_activity

=head1 Description

B<SWISS::CCcofactor> represents a comment on the topic 'CATALYTIC ACTIVITY'
within a Swiss-Prot or TrEMBL entry as specified in the user manual
http://www.expasy.org/sprot/userman.html .  Comments on other topics are stored
in other types of objects, such as SWISS::CC (see SWISS::CCs for more information).

Collectively, comments of all types are stored within a SWISS::CCs container
object.

=head1 Inherits from

SWISS::BaseClass

=head1 Attributes

=over

=item topic

The topic of this comment ('CATALYTIC ACTIVITY').

=item comment

The "text" version of this comment (without evidences and new lines).


=back
=head1 Methods

=item activity_rhea

The rhea id (e.g. RHEA:18037) of the reaction (only for new structured CC CATALYTIC ACTIVITY)

=item activity_rhea( $new_rhea_id, $new_optional_reaction_ev )

Set rhea activity to $new_rhea_id (and reaction ev to $new_optional_reaction_ev if provided)
(only for new structured CC CATALYTIC ACTIVITY)

=item activity_reaction

The reaction txt (e.g. agmatine + H2O = N-carbamoylputrescine + NH4(+))

=item activity_reaction( $new_reaction_txt )

Set reaction txt to $new_reaction_txt
(on old unstructured format: extra $ev could be added)

=item activity_ec

The reaction EC (e.g. 3.5.3.12)

=item activity_ec( $new_reaction_ec, , $new_optional_reaction_ev )

Set reaction EC to $new_reaction_ec (and reaction ev to $new_optional_reaction_ev if provided)
(only for new structured CC CATALYTIC ACTIVITY)

=item activity_participant

The reaction participants as a String Array (e.g. [ "ChEBI:CHEBI:15377", "ChEBI:CHEBI:28938", "ChEBI:CHEBI:58145", "ChEBI:CHEBI:58318" ])
(participant are shown as dbtype:id e.g. in "ChEBI:CHEBI:15377" CheBI is db type, CHEBI:15377 is (CheBi) id)
(only for new structured CC CATALYTIC ACTIVITY)

=item activity_participant( $new_participant_array )

Set reaction participants to $new_participant_array (array of strings)
(only for new structured CC CATALYTIC ACTIVITY)


=item directions

The directions: Array of direction structures as hash e.g. [ { 'txt' => 'left-to-right', 'rhea' => 'RHEA:16846', 'ev' => '{ECO:0000269|PubMed:29420286}' } ]
p.s. in well formed UniProtKB data 'txt' will be 'right-to-left' or 'left-to-right' (in that order); there should be no more than 2 directions (minimum: 0)

=item direction( $new_directions )

Set directions Array to $new_directions ...
(only for new structured CC CATALYTIC ACTIVITY)

=head2 Standard methods

=over

=item new

=item fromText

=item toString

Returns a string representation of this comment.

=back
