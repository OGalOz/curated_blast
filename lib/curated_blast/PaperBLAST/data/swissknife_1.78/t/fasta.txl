ID   DCTQ_RHOSH              Unreviewed;       227 AA.
AC   Q9LBD9;
DT   01-OCT-2000 (TrEMBLrel. 15, Created)
DT   01-OCT-2000 (TrEMBLrel. 15, Last sequence update)
DT   01-OCT-2000 (TrEMBLrel. 15, Last annotation update)
DE   Putative small C4-dicarboxylate integral membrane transport protein (Yadda).
GN   DCTQ.
OS   Rhodobacter sphaeroides (Rhodopseudomonas sphaeroides).
OC   Bacteria; Proteobacteria; alpha subdivision; Rhodobacter group;
OC   Rhodobacter.
OX   NCBI_TaxID=1063;
RN   [1]
RP   SEQUENCE FROM N.A.
RC   STRAIN=4P1;
RA   Omrani M.D.;
RL   Submitted (MAY-1997) to the EMBL/GenBank/DDBJ databases.
CC   -!- FUNCTION: REQUIRED FOR C4-DICARBOXYLATE TRANSPORT (BY SIMILARITY).
CC   -!- SUBCELLULAR LOCATION: INTEGRAL MEMBRANE PROTEIN (BY SIMILARITY).
SQ   SEQUENCE   227 AA;  24991 MW;  2208AD920C1230DA CRC64;
     MMRLLDRLEE TLIASLIAAA TGLIFVSVVQ RYSLGLLADG VAFFRGHDMP ELSAMMRSAY
     LGLREFNLVW AQELCIILFV WMAKFGAAYG VRTGIHVGID VLINKLDERK RGFFILLGLG
     AGALFTGIIA TLGGNFVWHM AQTSAISPDL ELPMWLVYLA IPLGSALMCF RFLQVAVIFA
     RTGELAHHDH GHVEGVDTED EGIDVLGSTF LKSPLTPRDL VEKPKDE
//
ID   DCTQ_RHOSH              Unreviewed;       227 AA.
AC   Q9LBD9;
DT   01-OCT-2000 (TrEMBLrel. 15, Created)
DT   01-OCT-2000 (TrEMBLrel. 15, Last sequence update)
DT   01-OCT-2000 (TrEMBLrel. 15, Last annotation update)
DE   Putative small C4-dicarboxylate integral membrane transport protein
DE   with a very very very very very very very very very very very very
DE   very very very very very very very very very very very very very
DE   very very very very very very very very very very very long name (Fragment).
OS   Rhodobacter sphaeroides (Rhodopseudomonas sphaeroides).
OC   Bacteria; Proteobacteria; alpha subdivision; Rhodobacter group;
OC   Rhodobacter.
OX   NCBI_TaxID=1063;
SQ   SEQUENCE   227 AA;  24991 MW;  2208AD920C1230DA CRC64;
     MMRLLDRLEE TLIASLIAAA TGLIFVSVVQ RYSLGLLADG VAFFRGHDMP ELSAMMRSAY
     LGLREFNLVW AQELCIILFV WMAKFGAAYG VRTGIHVGID VLINKLDERK RGFFILLGLG
     AGALFTGIIA TLGGNFVWHM AQTSAISPDL ELPMWLVYLA IPLGSALMCF RFLQVAVIFA
     RTGELAHHDH GHVEGVDTED EGIDVLGSTF LKSPLTPRDL VEKPKDE
//
ID   3MGH_SYNJB              Reviewed;         192 AA.
AC   Q2JI31;
DT   12-DEC-2006, integrated into UniProtKB/Swiss-Prot.
DT   07-MAR-2006, sequence version 1.
DT   08-FEB-2011, entry version 31.
DE   RecName: Full=Putative 3-methyladenine DNA glycosylase;
DE            EC=3.2.2.-;
GN   OrderedLocusNames=CYB_2817;
OS   Synechococcus sp. (strain JA-2-3B'a(2-13)) (Cyanobacteria bacterium
OS   Yellowstone B-Prime).
SQ   SEQUENCE   192 AA;  21330 MW;  71CAF20F5D810045 CRC64;
     MEWLSQPAPL VAPALLGMVL VRQFADGLQV RAQIVETEAY TAGDPACHAY RRKTQRNQVM
     FGPPGHLYIY RIYGLYHCLN IVTEPEGIPA AVLIRAAQLD RLPDWIPANK QNQPARAAAG
     PGLLCQALRI DGSHNGWRLE RAEAGQEGIW LEGSPSWQTQ LSIVQTTRIG ITQGAEIPWR
     WYIGGHPAVS RY
//
