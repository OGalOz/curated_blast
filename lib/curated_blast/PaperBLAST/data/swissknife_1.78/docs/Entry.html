<?xml version="1.0" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title>SWISS::Entry</title>
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:operations@isb-sib.ch" />
</head>

<body style="background-color: white">


<!-- INDEX BEGIN -->
<div name="index">
<p><a name="__index__"></a></p>

<ul>

	<li><a href="#name">Name</a></li>
	<li><a href="#description">Description</a></li>
	<li><a href="#example">Example</a></li>
	<li><a href="#attributes">Attributes</a></li>
	<li><a href="#methods">Methods</a></li>
	<ul>

		<li><a href="#data_access_methods">Data access methods</a></li>
	</ul>

</ul>

<hr name="index" />
</div>
<!-- INDEX END -->

<p>
</p>
<h1><a name="name">Name</a></h1>
<p>SWISS::Entry</p>
<p>
</p>
<hr />
<h1><a name="description">Description</a></h1>
<p>Main module to handle SWISS-PROT entries. One Entry object represents one SWISS-PROT entry and provides an API for its modification.</p>
<p>The basic concept is the idea of lazy parsing. If an Entry object is created from the entry in flat file format, the text is simply stored in the private text attribute of the entry object. The member objects of the entry are only created if they are dereferenced.</p>
<p>
</p>
<hr />
<h1><a name="example">Example</a></h1>
<PRE>
use SWISS::Entry;
# Read an entire record at a time
$/ = "\/\/\n";
while (<>){
  $entry = SWISS::Entry->fromText($_);
  print $entry->AC, "\n";
}
</pre><p>This minimum program reads entries from a file in SWISS-PROT format and prints the primary accession number for each of the entries.</p>
<p>
</p>
<hr />
<h1><a name="attributes">Attributes</a></h1>
<p>The following attributes represent member objects. They can be accessed like e.g. $entry-&gt;IDs</p>
<dl>
<dt><strong><a name="ids" class="item">IDs</a></strong></dt>

<dd>
<p>ID line object</p>
</dd>
<dt><strong><a name="acs" class="item">ACs</a></strong></dt>

<dt><strong><a name="dts" class="item">DTs</a></strong></dt>

<dt><strong><a name="des" class="item">DEs</a></strong></dt>

<dt><strong><a name="gns" class="item">GNs</a></strong></dt>

<dt><strong><a name="oss" class="item">OSs</a></strong></dt>

<dt><strong><a name="ocs" class="item">OCs</a></strong></dt>

<dt><strong><a name="refs" class="item">Refs</a></strong></dt>

<dd>
<p>The reference block object</p>
</dd>
<dt><strong><a name="ccs" class="item">CCs</a></strong></dt>

<dt><strong><a name="kws" class="item">KWs</a></strong></dt>

<dt><strong><a name="drs" class="item">DRs</a></strong></dt>

<dt><strong><a name="fts" class="item">FTs</a></strong></dt>

<dt><strong><a name="stars" class="item">Stars</a></strong></dt>

<dd>
<p>Object for the annotator's section stored in the ** lines.</p>
</dd>
<dt><strong><a name="sqs" class="item">SQs</a></strong></dt>

<dd>
<p>The sequence object.</p>
</dd>
</dl>
<p>
</p>
<hr />
<h1><a name="methods">Methods</a></h1>
<dl>
<dt><strong><a name="new" class="item">new</a></strong></dt>

<dd>
<p>Return a new Entry object</p>
</dd>
<dt><strong><a name="initialize" class="item">initialize</a></strong></dt>

<dd>
<p>Initialise an Entry object and return it.</p>
</dd>
<dt><strong><a name="update_force" class="item">update [force]</a></strong></dt>

<dd>
<p>Update an
entry. The content of the member objects is written back into the private text
attribute of the entry if necessary. If $force is true, an update of all 
member objects is forced.</p>
</dd>
<dt><strong><a name="reformat" class="item">reformat</a></strong></dt>

<dd>
<p>Reformat all fields of an entry.</p>
</dd>
<dt><strong><a name="fromtext_text_fullparse_removeinternalcomments" class="item">fromText $text [, $fullParse[, $removeInternalComments]]</a></strong></dt>

<dd>
<p>Create an Entry object from the text $text. If $fullParse is true, the entry is
parsed at creation time. Otherwise the individual line objects are only 
created if they are dereferenced. If $removeInternalComments is true, wild comments
and indentation will be removed from the text before the parsing is done. [NOTE: 
wild comments are lines starting with a double asterisk located outside the Stars
section, and indented lines are lines starting with spaces. Both are used internally
by SWISS-PROT annotators during their work and excluded from internal and external
releases.]</p>
</dd>
<dt><strong><a name="totext_insertinternalcomments" class="item">toText [$insertInternalComments]</a></strong></dt>

<dd>
<p>Return the entry in flat file text format. If internal comments and indentation
have been removed as specified in the parameters to <code>fromText()</code>, you may wish
to reinsert them in the text output by setting $insertInternalComments to true.</p>
</dd>
<dt><strong><a name="tofasta" class="item">toFasta</a></strong></dt>

<dd>
<p>Return the entry in Fasta format.</p>
</dd>
<dt><strong><a name="equal" class="item">equal</a></strong></dt>

<dd>
<p>Returns True if two entries are equal, False otherwise</p>
</dd>
</dl>
<p>The following methods are provided for your convenience. They are shortcuts for methods of the individual line objects.</p>
<dl>
<dt><strong><a name="id" class="item">ID</a></strong></dt>

<dd>
<p>Returns the primary ID of the entry.</p>
</dd>
<dt><strong><a name="ac" class="item">AC</a></strong></dt>

<dd>
<p>Returns the primary AC of the entry.</p>
</dd>
<dt><strong><a name="sq" class="item">SQ</a></strong></dt>

<dd>
<p>Returns the sequence of the entry.</p>
</dd>
<dt><strong><a name="ev" class="item">EV</a></strong></dt>

<dd>
<p>Returns the EV (evidence) object of an entry. SWISS-PROT internal method.</p>
</dd>
</dl>
<p>
</p>
<h2><a name="data_access_methods">Data access methods</a></h2>
<dl>
<dt><strong><a name="text" class="item">text</a></strong></dt>

<dd>
<p>Returns the current text of the entry. 
<strong>Quick and dirty!</strong> No update of the text is performed before.</p>
</dd>
<dt><strong><a name="database_code" class="item">database_code</a></strong></dt>

<dd>
<p>Is it a SWISS-PROT, TREMBL or TREMBLNEW entry?
database_code tries to find it out.
Return values are S for SWISS-PROT, 3 for TREMBL, Q for TREMBLNEW, ? for unknown.</p>
</dd>
<dt><strong><a name="isfragment" class="item">isFragment</a></strong></dt>

<dd>
<p>Returns true if the DE line indicates a fragment, or of the entry 
contains a NON_CONS or NON_TER feature.</p>
</dd>
<dt><strong><a name="iscurated" class="item">isCurated</a></strong></dt>

<dd>
<p>Returns 1 if the entry is a curated entry, 
0 otherwise.</p>
<p>SWISS-PROT internal use only.</p>
</dd>
</dl>

</body>

</html>
