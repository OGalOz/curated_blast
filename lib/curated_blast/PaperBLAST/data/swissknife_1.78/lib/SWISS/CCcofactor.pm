package SWISS::CCcofactor;

use vars qw($AUTOLOAD @ISA %fields);

use Carp;
use strict;
use SWISS::TextFunc;
use SWISS::ListBase;
use SWISS::CC;


BEGIN {
  @ISA = ( 'SWISS::ListBase' );
  
  %fields = (
          form        => undef,  
	      note        => undef, # [ txt, ev ]
	      note_blocks => undef # [ [txt, ev ] ]
	    );
}


sub new {
  my $ref = shift;
  my $class = ref($ref) || $ref;
  my $self = new SWISS::ListBase;
  $self->rebless($class);
  return $self;
}


sub fromText {
    my $class = shift;
    my $textRef = shift;
    my $self = new SWISS::CCcofactor;
    my $text = $$textRef;
    $self->initialize();
  
    $text =~ s/ *-!- COFACTOR: +//;
    $text =~ s/; {2,}/; /g;
    $text =~ s/, {2,}/, /g;
    
    if ( $text !~ /Name=/ && $text !~ /Note=/ ) { # old format!
    	  my $note = $text;
    	( my $note_ev = $1 ) if $note =~ s/($SWISS::TextFunc::evidencePattern)\.?//;
    	$self->{ note } = [ $note.".", $note_ev ] if $note;
    }
    else { # new format
        ( my $form    = $1 )           if $text =~ s/^([^:]+): (?=N)//;
        ( my $note    = $1 ) =~ s/;$// if $text =~ s/ ?Note=(.+)$//;
        $self->{ form }  = $form if $form;
        $self->{ note_blocks } = SWISS::CC::parse2Blocks( $note ) if $note; #[ $note, $note_ev ] if $note;
        foreach my $name ( split / +(?=Name=)/, $text ) {
            my $ev = $name =~ s/ +Evidence=($SWISS::TextFunc::evidencePattern);// ? " ".$1 : undef;
            $self->add( [ $name, $ev ] );
        } # $name = e.g. "Name=[2Fe-2S] cluster; Xref=ChEBI:CHEBI:49601;" includes chebi xref! not further parsed!
    }

    $self->{_dirty} = 0;
    return $self;
}


sub topic {
	return "COFACTOR";
}


sub toString {
    my ( $self ) = @_;
    
    my $form = $self->{form} ? " $self->{form}:" : "";
    my $text = "CC   -!- COFACTOR:$form\n";
    
    foreach my $name_ev ( @{ $self->{list} } ) {
        my ( $name, $ev ) = @$name_ev;
        my   $line = $name;
        if ( $ev ) { $ev=~s/^ +//; $line .= " Evidence=".$ev.";" }
        $text .= SWISS::TextFunc->wrapOn( 'CC       ',"CC         ", $SWISS::TextFunc::lineLength, $line, "(?<=;) ", "(?<=,) ", $SWISS::TextFunc::textWrapPattern1 );
    }
    if ( $self->{note} ) { # old format
     	my ( $note, $note_ev ) = @{ $self->{ note } };
        $note_ev ||= "";
        $note      = "Note=" . $note . $note_ev . ";";
        $text     .= SWISS::TextFunc->wrapOn( 'CC       ',"CC       ", $SWISS::TextFunc::lineLength, $note );
    }
    if ( $self->{note_blocks} ) {
        my $note  = "Note=" . SWISS::CC::blocks2String( $self->{ note_blocks }, "" ) . ";";
        $text    .= SWISS::TextFunc->wrapOn( 'CC       ',"CC       ", $SWISS::TextFunc::lineLength, $note );
    }
    
    return $text; 
}


sub comment {
    my ( $self, $with_ev ) = @_;
    
    my $text = "";
    foreach my $name_ev ( @{ $self->{list} } ) {
        my ( $name, $ev ) = @$name_ev;
        $text .= ( $text ? " " : "" ) . $name;
        if ( $ev && $with_ev ) { $ev=~s/^ +//; $text .= " Evidence=".$ev.";" }
    }
    if ( $self->{ note } ) {
    	my ( $note, $note_ev ) = @{ $self->{ note } };
    	$note_ev = "" unless $with_ev && $note_ev;
        $note      = "Note=" . $note . $note_ev . ";";
        $text .= ( $text ? " " : "" ) . $note;
    }
    
    return $text;
}


sub structured_names {
    my ( $self, $values ) = @_;

    if ( defined $values ) { # set
        $self->{list} = \ map { [ "Name=".$_->{name}."; Xref=ChEBI:".$_->{chebi}, $_->{ev} ] } @{ $values };
    }
    my @out = map { # get
        my $elem = {};
        $elem->{ev} = $_->[1];                                    # e.g. ECO:0000255|HAMAP-Rule:MF_01841
        my @vals = grep { $_ ne '' } split /Name=|Xref=/, $_->[0];
        ( $elem->{name}  = $vals[0] || "" ) =~ s/; ?$//;          # e.g. FAD
        ( $elem->{chebi} = $vals[1] || "" ) =~ s/^ChEBI:|; ?$//g; # e.g. CHEBI:57692
        print "----STRUCT COF-->[".$elem->{name}."]\n";
        $elem;
    } @{ $self->{list} };
    return \@out;
} # returns Array of { name->..., chebi->..., ev->... }


#sub sort {
#    my $self = shift;
#    $self->{ list } = sort { lc( $a->[0] ) cmp lc( $b->[0] ) } @{ $self->{ list } };
#}


1;

__END__

=head1 Name

SWISS::CCcofactor

=head1 Description

B<SWISS::CCcofactor> represents a comment on the topic 'COFACTOR'
within a Swiss-Prot or TrEMBL entry as specified in the user manual
http://www.expasy.org/sprot/userman.html .  Comments on other topics are stored
in other types of objects, such as SWISS::CC (see SWISS::CCs for more information).

Collectively, comments of all types are stored within a SWISS::CCs container
object.

=head1 Inherits from

SWISS::ListBase.pm

=head1 Attributes

=over

=item topic

The topic of this comment ('COFACTOR').

=item comment

The "text" version of this comment (without evidences and new lines).

=item note

The note and evidence (Note= in new format or full description in old format)
reference to an array of [ $note, $note_ev ] (strings)

=item elements

An array of [name_str, evidence_tags_str], if any.

=back
=head1 Methods

=item structured_names

(only for new structured CC COFACTOR)
The names as array of 'name' txt, 'chebi' id, 'ev' string (hash) structures
e.g. [ { 'name'->'FAD', 'chebi'->'CHEBI:57692', 'ev'->'ECO:0000250' } ]

=item structured_names( $new_values )

(only for new structured CC COFACTOR)
Set names with array of 'name' txt, 'chebi' id, 'ev' string (hash) structures like returned by get-er version


=head2 Standard methods

=over

=item new

=item fromText

=item toString

Returns a string representation of this comment.

=back
