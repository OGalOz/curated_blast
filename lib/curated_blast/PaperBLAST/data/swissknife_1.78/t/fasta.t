# Test for FASTA output

use Carp;

# * Test loading
BEGIN { 
  $| = 1; print "1..2\n"; 
  use vars qw($loaded); 
  $^W = 0;
}

END { print "not ok 1\n" unless $loaded; }

$loaded = 1;
print "ok 1\n";    # 1st test passes.

sub test ($$;$) {
    my($num, $true,$msg) = @_;
    print($true ? "ok $num\n" : "not ok $num $msg\n");
}

my $where = -d 't' ? "t/" : "";
my $testin = "${where}fasta.txl";
my $testout  = "${where}fasta.txl.out";
my $expectedout = "${where}fasta.txl.expected";

open (IN, $testin);
open (OUT, ">$testout");

use SWISS::Entry;
use Data::Dumper;

# Read an entire record at a time
$/ = "\/\/\n";
 
while (<IN>){
  $entry = SWISS::Entry->fromText($_);
  print OUT $entry->toFasta();
}

close IN;
close OUT;
print "checking expected output...\n";

test 2, system('diff', $testout, $expectedout) == 0, "diff $testout $expectedout";
