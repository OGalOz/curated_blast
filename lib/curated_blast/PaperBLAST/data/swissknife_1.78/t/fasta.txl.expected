>tr|Q9LBD9|DCTQ_RHOSH Putative small C4-dicarboxylate integral membrane transport protein OS=Rhodobacter sphaeroides OX=1063 GN=DCTQ
MMRLLDRLEETLIASLIAAATGLIFVSVVQRYSLGLLADGVAFFRGHDMPELSAMMRSAY
LGLREFNLVWAQELCIILFVWMAKFGAAYGVRTGIHVGIDVLINKLDERKRGFFILLGLG
AGALFTGIIATLGGNFVWHMAQTSAISPDLELPMWLVYLAIPLGSALMCFRFLQVAVIFA
RTGELAHHDHGHVEGVDTEDEGIDVLGSTFLKSPLTPRDLVEKPKDE
>tr|Q9LBD9|DCTQ_RHOSH Putative small C4-dicarboxylate integral membrane transport protein with a very very very very very very very very very very very very very very very very very very very very very very very very very very very very very very very very very very very very long name (Fragment) OS=Rhodobacter sphaeroides OX=1063
MMRLLDRLEETLIASLIAAATGLIFVSVVQRYSLGLLADGVAFFRGHDMPELSAMMRSAY
LGLREFNLVWAQELCIILFVWMAKFGAAYGVRTGIHVGIDVLINKLDERKRGFFILLGLG
AGALFTGIIATLGGNFVWHMAQTSAISPDLELPMWLVYLAIPLGSALMCFRFLQVAVIFA
RTGELAHHDHGHVEGVDTEDEGIDVLGSTFLKSPLTPRDLVEKPKDE
>sp|Q2JI31|3MGH_SYNJB Putative 3-methyladenine DNA glycosylase OS=Synechococcus sp. (strain JA-2-3B'a(2-13)) OX=? GN=CYB_2817 SV=1
MEWLSQPAPLVAPALLGMVLVRQFADGLQVRAQIVETEAYTAGDPACHAYRRKTQRNQVM
FGPPGHLYIYRIYGLYHCLNIVTEPEGIPAAVLIRAAQLDRLPDWIPANKQNQPARAAAG
PGLLCQALRIDGSHNGWRLERAEAGQEGIWLEGSPSWQTQLSIVQTTRIGITQGAEIPWR
WYIGGHPAVSRY
